### PR TITLE
feat: [CDS-46699]: Add download directory methods to AwsS3HelperServi…

### PR DESCRIPTION
**Pull request by akhilpandey1**
…ceDelegate

## Describe your changes

As of now, we don't have a method to download remote directory from AWS S3, I've created and implemented a new AWS S3 TransferManager API for this

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40420)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java [hunk 1]
```diff
@@ -159,6 +163,56 @@ public S3Object getObjectFromS3(
     return null;
   }
 
+  public boolean downloadS3Directory(AwsConfig awsConfig, String bucketName, String key, File workingDirectory)
+      throws InterruptedException {
+    TransferManager transferManager =
+        TransferManagerBuilder.standard().withS3Client(getAmazonS3Client(awsConfig)).build();
+    MultipleFileDownload download = transferManager.downloadDirectory(bucketName, key, workingDirectory);
+    try {
+      download.waitForCompletion();
```
Comment (vlica-harness) [inline 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java:172]: waitForComplection can throw AmazonClientException, AmazonServiceException, InterruptedException.
In this class implementation we have the handler logic for AmazonClientException, AmazonServiceException.
Why not to follow the same process?

File: 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java [hunk 2]
```diff
@@ -159,6 +163,56 @@ public S3Object getObjectFromS3(
...
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new InterruptedException(ExceptionMessageSanitizer.sanitizeException(e).getMessage());
+    }
+    return true;
```
Comment (vlica-harness) [inline 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java:177]: shoudn't you use **transferManager.shutdownNow()** before finishing this method?

File: 930-delegate-tasks/src/main/java/software/wings/service/intfc/aws/delegate/AwsS3HelperServiceDelegate.java [hunk 3]
```diff
@@ -34,4 +35,8 @@ ObjectMetadata getObjectMetadataFromS3(
       AwsConfig awsConfig, List<EncryptedDataDetail> encryptionDetails, String bucketName, String key);
   S3Object getObjectFromS3(
       AwsConfig awsConfig, List<EncryptedDataDetail> encryptionDetails, String bucketName, String key);
+  boolean downloadS3Directory(AwsConfig awsConfig, String bucketName, String key, File workingDirectory)
+      throws InterruptedException;
+  boolean downloadS3DirectoryUsingS3URI(AwsConfig awsConfig, String s3URI, File workingDirectory)
+      throws InterruptedException;
```
Comment (vlica-harness) [inline 930-delegate-tasks/src/main/java/software/wings/service/intfc/aws/delegate/AwsS3HelperServiceDelegate.java:41]: Not sure the methods need to have throws in the definition in our case, do we  want to mandate the clients of the interface to handle differently this exception or do we expect this?

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40420)* status: 2 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 166 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHTd1uq9J8lww_F4awo:-NHTd1uq9J8lww_F4awp:b-n5pdkb) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L166)):*
> ```Java
>   }
> 
>   public boolean downloadS3Directory(AwsConfig awsConfig, String bucketName, String key, File workingDirectory)
> ```

I would call it destinationDirectory.
___
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 189 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHTcEH5Ahpc9QLpzJjz:-NHTcEH5Ahpc9QLpzJk-:bilfno4) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L189)):*
> ```Java
>    *
>    */
>   public boolean downloadS3DirectoryUsingS3URI(AwsConfig awsConfig, String s3URI, File workingDirectory)
> ```

I think the method above (which takes bucketName and key)  is enough. It's up to the users of that service to resolve them.

**Comment by akhilpandey1** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 177 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHOY6COTkooeSACw1aa-r7-177:-NHTeyAb3TTFzt1w-isK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L177)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

shoudn't you use **transferManager.shutdownNow()** before finishing this method? 
 
</blockquote></details>

Done.

**Comment by akhilpandey1** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 166 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHTd1uq9J8lww_F4awo:-NHTiEtf7V2aAvxBBgTT:bejoo3b) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L166)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

I would call it destinationDirectory.
</blockquote></details>

makes sense, sure

**Comment by akhilpandey1** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 172 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHOY6COTkooeSACw1aa-r7-172:-NHTiJ0i3VFB-g-0VrLY:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L172)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

waitForComplection can throw AmazonClientException, AmazonServiceException, InterruptedException.
In this class implementation we have the handler logic for AmazonClientException, AmazonServiceException.
Why not to follow the same process? 
</blockquote></details>

Done

**Comment by tmacari-harness** [comment]
Also let's add unit test for new method

**Comment by akhilpandey1** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java` line 189 at r7](https://reviewable.io/reviews/harness/harness-core/40420#-NHTcEH5Ahpc9QLpzJjz:-NHUFdGD1w0TZh7C4hjm:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ca4eb6e51ee04d2cd9adbae66e5d5a6cfeb7c003/930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsS3HelperServiceDelegateImpl.java#L189)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

I think the method above (which takes bucketName and key)  is enough. It's up to the users of that service to resolve them.
</blockquote></details>

Done.

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40420)* status: 3. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40420)* status: 3. unresolved discussions.

**Comment by akhilpandey1** [comment]
Done

**Review by akhilpandey1 — COMMENTED** [review]
> Reviewed 1 of 2 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40420)* status: 3. unresolved discussions.

**Pull request merged**